### ISSUE TITLE
Possible data formatting issue affecting reproduction of PRISM results

### ISSUE THREAD

**Issue by ym689** [facebookresearch/LoRe#2]
Hi, thanks for releasing this repository.

When trying to reproduce the reported results, I found that the accuracies are consistently close to 1.0. After checking the code, I noticed a potential data formatting issue that may significantly affect model training and evaluation.

In the data construction code, `chosen_utterance` and `rejected_utterance` are handled inconsistently:

```python
# the single resporse string
chosen_utterance = turn["chosen_utterance"][0]
entry['extra_info']['chosen_utterance'] = chosen_utterance

# all rejected response strings
rejected_utterance = turn["rejected_utterance"]
entry['extra_info']['rejected_utterance'] = rejected_utterance
```
Later, they are both wrapped as assistant messages:
```python
chosen = [{"content": entry["extra_info"]["chosen_utterance"], "role": "assistant"}]
rejected = [{"content": entry["extra_info"]["rejected_utterance"], "role": "assistant"}]
```
As a result, chosen_utterance is a string, while rejected_utterance is a list of strings.
When computing embeddings or feeding the data into the model, rejected_utterance is implicitly converted into a string representation of a list (e.g., `"['...']"`), which always contains the tokens `[` and `]`.

This introduces a trivial and spurious feature that the model can easily exploit to distinguish rejected samples from chosen ones, without learning any meaningful semantic preference. In practice, this may explain why the accuracy is consistently very high (often close to 1.0).

I hope my understanding is helpful. Please feel free to let me know if I have misunderstood anything or if there is additional context I might be missing. Thanks again for sharing your work.

**Comment by Avinandan22** [comment]
Hi!

Thanks for carefully going through the code and finding the bug! I realized this is only an issue with the PRISM data preparation, but not present in the other datasets considered. This bug got introduced while fixing the formating for multi-turn conversations. Attaching a snippet of code from initial commit that could help you fix this.

Do you mind running experiments by treated them as different pairs of examples (I believe for each chosen response, there are 3 rejected pairs), and see what numbers you are getting? 

If it works out, feel free to raise a pull request and I'll merge the changes.

`for rejected in turn["rejected_utterance"]:
                        # skip too long examples
                        if len(turn["user_utterance"][0]) + len(rejected) > max_text_length:
                            continue
                        # new_examples["prompt"].append(prompt)
                        # new_examples["chosen"].append(turn['chosen_utterance'][0])
                        # new_examples["rejected"].append(rejected)
                        new_examples[user_id][dialog_id]["prompt"].append(prompt)
                        new_examples[user_id][dialog_id]["chosen"].append(turn['chosen_utterance'][0])
                        new_examples[user_id][dialog_id]["rejected"].append(rejected)`

Thanks!

**Comment by ym689** [comment]
Hi, thanks for the clarification.

I’v run the experiment where each rejected response is paired with the chosen response as a separate training sample (i.e., treating them as different preference pairs). Note that not every turn contains three rejected responses.

Below are the resulting dataset statistics after restructuring:

| Split                | # Samples |
|----------------------|-----------|
| Seen user – Train    | 9,913     |
| Seen user – Test     | 9,932     |
| Unseen user – Train  | 9,725     |
| Unseen user – Test   | 9,795     |

I ran experiments under two settings, K = 25 and K = 50, and the corresponding results are shown in the attached figure.

The numbers are slightly different from those reported in the paper, which may be due to differences in hyperparameter settings or other implementation details. Please feel free to point out if there is anything I should adjust or double-check.

If everything looks reasonable, I’m happy to clean up the code and submit a pull request.

Thanks again for your help and for maintaining the repository.

<img width="2153" height="1407" alt="Image" src="https://github.com/user-attachments/assets/a3ed4387-ca58-4845-994a-2317d00b5545" />

**Comment by Avinandan22** [comment]
Thanks for the quick changes. I'm guessing the hyperparameters would change quite a bit. 

In train_basis.py, can you set the alpha_list = [0.0], and also when reporting the plots, can you have rank = 1 too, so that it is easier to see how much better it does from the BT model.
